Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -583,9 +583,75 @@ void GraphicsContextSkia::drawLinesForText(const FloatPoint& point, float thickn
         fillRect(dash, localStrokeColor);
 }
 
-void GraphicsContextSkia::drawDotsForDocumentMarker(const FloatRect&, DocumentMarkerLineStyle)
+void GraphicsContextSkia::drawDotsForDocumentMarker(const FloatRect& boundaries, DocumentMarkerLineStyle style)
 {
-    notImplemented();
+    if (style.mode != DocumentMarkerLineStyleMode::Spelling
+        && style.mode != DocumentMarkerLineStyleMode::Grammar)
+        return;
+
+    canvas().drawPath(createFillableTriangleWave(boundaries), createFillPaint(static_cast<SkColor>(style.color)));
+}
+
+// Creates a path comprising of two triangle waves separated by some empty space in Y axis.
+// The empty space can be filled using SkPaint::kFill_Style thus forming an elegant triangle wave.
+// Such triangle wave can be used e.g. as an error underline for text.
+SkPath GraphicsContextSkia::createFillableTriangleWave(const FloatRect& boundaries)
+{
+    SkPath path;
+
+    double x = boundaries.x();
+    const double y = boundaries.y();
+    double width = boundaries.width();
+    const double height = boundaries.height();
+    static const double heightSquares = 2.5;
+
+    const double square = height / heightSquares;
+    const double halfSquare = 0.5 * square;
+
+    const double unitWidth = (heightSquares - 1.0) * square;
+    const int widthUnits = static_cast<int>((width + 0.5 * unitWidth) / unitWidth);
+
+    x += 0.5 * (width - widthUnits * unitWidth);
```

Comment:
This could be `double x = boundaries.x() + (0.5 * (width - widthUnits * unitWidth));` since `x` is not used before this.